### ISSUE TITLE
poll_read_priv unconditionally clear_ready!()

### ISSUE THREAD

**Issue by isomer** [jbaublitz/neli#166]
**Version of neli**
0.6.0

**Does the documentation specify that neil's behavior is correct?**
No - tokio says you only should clear the ready signal when the socket would block.

**Does there appear to be a mistake in the netlink documentation based on the kernel code?**
This is independent of netlink, it's a misuse of the tokio apis.

**Describe the bug**

neli::socket::tokio::NlSocket uses an AsyncFd to manage the events on the socket:

```rust
    fn poll_read_priv(
        async_fd: &AsyncFd<super::NlSocket>,
        cx: &mut Context,
        buf: &mut ReadBuf,
    ) -> Poll<io::Result<usize>> {
        let mut guard = ready!(async_fd.poll_read_ready(cx))?;
        guard.clear_ready();
        let socket = async_fd.get_ref();
        let bytes_read = socket.recv(buf.initialized_mut(), 0)?;
        buf.advance(bytes_read);
        Poll::Ready(Ok(bytes_read))
    }
```

This calls `let mut guard = ready!(async_fd.poll_read_ready(cx))?;`, but then unconditionally calls `guard.clear_ready();`.  The AsyncFd documentation for clear_ready() says:

> It is critical that this function not be called unless your code actually observes that the file descriptor is not ready. Do not call it simply because, for example, a read succeeded; it should be called when a read is observed to block.

Instead this should call `socket.recv()` and inspect to see if the error return is `WouldBlock` and only call `clear_ready()` then.  I think this can be better done with try_io.

eg (untested code):
```rust
    fn poll_read_priv(
        async_fd: &AsyncFd<super::NlSocket>,
        cx: &mut Context,
        buf: &mut ReadBuf,
    ) -> Poll<io::Result<usize>> {
     loop {
        let mut guard = ready!(async_fd.poll_read_ready(cx))?;
        match guard.try_io(|socket| socket.get_ref().recv(buf.initialized_mut(), 0)) {
          Ok(Ok(bytes)) => { buf.advance(bytes); return Poll::Ready(Ok(bytes)); }
          Ok(result) => return Poll::Ready(result),
          Err(_would_block) => continue,
        }
      }
  }
```

Without this, if there are more bytes than the buffer has waiting, then the program blocks waiting for the fd to become ready again, but since it's already ready, epoll() never returns the read event to tokio, blocking the program.

The write version of the function seems to have the same bug (although I don't think I've ever triggered it), with basically the same fix.

**To Reproduce**
1. Send a request onto a netlink socket that has a reply, do not wait for the reply.
2. Immediately send another request onto the netlink socket that also has a reply.
3. Attempt to wait for both replies.  Note that only the first reply arrives, and the second does not.

**Expected behavior**
neli should not block forever waiting for a ready signal that will never come.

**Comment by jbaublitz** [comment]
@isomer Thanks for reporting this! This misuse is largely due to a corner case that I bumped into in a separate project. In this project, we are forced to clear the readiness on each file descriptor readiness indicator as this file descriptor is not actually readable and returns an IO error when you attempt to do so. I talked with the tokio team and they agreed that in this corner case, you do have to clear the file descriptor each time or else it will cause the file descriptor to report readiness indefinitely. That however is not the case here and I must have forgotten to update it to the correct usage here after having that discussion. I'm putting up a PR now.
